# Задание 25. Обработка целочисленной информации

Задачи и разбор можно посмотреть на сайте [Полякова](https://kpolyakov.spb.ru/school/ege.htm) в этом [файле](https://kpolyakov.spb.ru/download/ege25.doc)

>В данном ноутбуке некоторые ячейки с кодом будут содержать чуть более сложный код, не нужный для решения задач этого номера. Он будет использоваться для демонстрации различных вещей, и нас будет интересовать только его результат выполнения. 

Решим вначале две общие задачи.

## Задача 1. Количество делителей натурального числа $n$

При такой постановке вопроса у нас есть всего один способ получения ответа &mdash; перебор возможных чисел с проверкой не является ли оно делителем $n$.

В простейшем случае мы можем перебрать все числа $d$ от $1$ до $n$:

```python
for d in range(1, n + 1):
```
>Для тех, кто слышал, что можно при решении этой задачи перебирать возможные делители не до $n$, а до какого-то меньшего числа, и, кажется, равного $\sqrt{n}$, обещаем, что это будет рассмотренно далее с возможными подводными камнями, которые возникают при использовании этого несомненно более эффективного перебора. 

А пока что для каждого числа от $1$ до $n$ проверим, является ли оно делителем числа $n$

```python
if n % d == 0:
  # d является делителем n
```

Для делителей будем увеличивать переменную-счетчик, которую предварительно надо обнулить:

```python
count = 0
for d in range(1, n + 1):
  if n % d == 0:
    # d является делителем n - увеличиваем счётчик
    count += 1 
```

Найдем количество делителей числа $n = 100$:

In [ ]:
n = 100
count = 0
for d in range(1, n + 1):
  if n % d == 0:
    count += 1
    
print(count)

9


Мы можем измерить время выполнения кода, если в первой строке ячейки с кодом добавим магическую команду `%%time`

In [ ]:
%%time
n = 100
count = 0
for d in range(1, n + 1):
  if n % d == 0:
    count += 1
    
print(count)

9
CPU times: user 1.19 ms, sys: 0 ns, total: 1.19 ms
Wall time: 1.21 ms


По строке `Wall time: 1.62 ms` можем сделать вывод, что время выполнения кода в этой ячейке составило $1.62$ миллисекунды.

Для разных запусков время может немного отличаться, но различия не будут слишком большими.

Посмотрим, сколько времени понадобится для обработки разных по величине чисел $n$. На сам код в следующей ячейке не смотрим, смотрим только на результат.

In [ ]:
import time

for k in range(3, 10):
  start_time = time.time()
  
  n = 10**k
  count = 0
  for d in range(1, n + 1):
    if n % d == 0:
      count += 1
  
  total_time = time.time() - start_time
  
  print(f'n=10^{k}:\tcount={count}\t\ttime: {round(total_time, 4)} sec')

n=10^3:	count=16		time: 0.0001 sec
n=10^4:	count=25		time: 0.0012 sec
n=10^5:	count=36		time: 0.0176 sec
n=10^6:	count=49		time: 0.1287 sec
n=10^7:	count=64		time: 1.2903 sec
n=10^8:	count=81		time: 12.7258 sec
n=10^9:	count=100		time: 125.4869 sec


Видим, что при возрастании числа $n$ в $10$ раз время тоже возрастает примерно в $10$ раз. Поэтому для чисел порядка $10$ миллиардов ($10^{10}$) мы можем ожидать, что время выполнения кода составит около трети часа ($\approx 1200$ секунд), а для $100$ миллиардов ($10^{11}$) &mdash; около $3$ часов.

Можно запомнить грубую оценку: за приемлемое время мы можем выполнить примерно миллиард ($10^9$) итераций. 

Тогда если нам понадобится находить количество делителей для, например, тысячи чисел, то чтобы общее количество итераций не превышало $10^9$ на одно число должно приходиться около $\frac{10^9}{10^3} = 10^6$, то есть миллиона итераций, и мы можем рассчитывать успеть обработать только числа, которые не превышают миллион.

Подобные оценки приблизительные, но позволяют грубо оценить сможем ли мы обойтись при решении задачи только таким простым способом.

In [ ]:
%%time
a, b = 174457, 174505
for n in range(a, b + 1):  
  count = 0
  for d in range(1, n + 1):
    if n % d == 0:
      count += 1
  if count == 4:
    print(n)

174459
174461
174463
174473
174479
174485
174497
174502
CPU times: user 1.13 s, sys: 3.47 ms, total: 1.13 s
Wall time: 1.14 s


In [ ]:
%%time
a, b = 174457, 174505
for n in range(a, b + 1):  
  count = 0
  divisors = []
  for d in range(1, n + 1):
    if n % d == 0:
      count += 1
      divisors.append(d)
  if count == 4:
    print(n, divisors)

174459 [1, 3, 58153, 174459]
174461 [1, 7, 24923, 174461]
174463 [1, 59, 2957, 174463]
174473 [1, 13, 13421, 174473]
174479 [1, 149, 1171, 174479]
174485 [1, 5, 34897, 174485]
174497 [1, 211, 827, 174497]
174502 [1, 2, 87251, 174502]
CPU times: user 1.15 s, sys: 4.92 ms, total: 1.16 s
Wall time: 1.17 s


In [ ]:
%%time
a, b = 174457, 174505
for n in range(a, b + 1):  
  count = 0
  divisors = []
  for d in range(1, n + 1):
    if n % d == 0:
      count += 1
      divisors.append(d)
  if count == 4:
    print(*divisors[1:3])

3 58153
7 24923
59 2957
13 13421
149 1171
5 34897
211 827
2 87251
CPU times: user 1.13 s, sys: 2.92 ms, total: 1.14 s
Wall time: 1.16 s


In [ ]:
%%time
a, b = 174457, 174505
for n in range(a, b + 1):  
  divisors = []
  for d in range(1, n + 1):
    if n % d == 0:
      divisors.append(d)
  if len(divisors) == 4:
    print(*divisors[1:3])

3 58153
7 24923
59 2957
13 13421
149 1171
5 34897
211 827
2 87251
CPU times: user 1.11 s, sys: 4.12 ms, total: 1.11 s
Wall time: 1.13 s


## Задача 2. Является ли натуральное число $n$ простым

Эту задачу можно свести к предыдущей, проверив в конце, что переменная `count` равна $2$ (два делителя - это $1$ и $n$):

```python
count = 0
for d in range(1, n + 1):
  if n % d == 0:
    count += 1 
if count == 2:
  # число n - простое
  ...
```

Но можно немного оптимизировать. Для ответа на вопрос задачи нам нет необходимости находить количество всех делителей. Если мы нашли хотя бы один, отличный от $1$ и $n$, то число уже будет составным.

>Кстати, делители числа $n$, равные $1$ и $n$, называются *тривиальными*

Чтобы исключить из перебора тривиальные делители $1$ и $n$, изменим цикл:

```python
for d in range(2, n):
  ...
```

И добавим прерывание цикла сразу после обнаружени первого делителя:

```python
...
  if n % d == 0:
    count += 1
    break
```

Тогда если `count` осталась равной $0$, то число будет простым:

```python
...
if count == 0:
  # число n - простое
  ...
```

Соберем всё вместе:

In [ ]:
n = 100
count = 0
for d in range(2, n):
  if n % d == 0:
    count += 1
    break

if count == 0:
  print(n, '- простое')
else:
  print(n, '- составное')

100 - составное


Для использования в последующих задачах будет весьма удобно реализовать вышеприведённый код в виде функции `is_prime(n)`, которая будет возвращать логическое значение, являющееся ответом на вопрос, являестя ли число $n$ простым?

В этом случае нам не понадобится даже переменная `count`, так как при обнаружении первого делителя можно сразу вернуть `False`.

И, так как при возврате значения из функции цикл всё равно не будет выполняться дальше, то нет необходимости в операторе `break`.

В том случае, если цикл закончился, а возврата из функции не произошло, то значит, у числа $n$ не нашлось ни одного нетривиалного делителя, то есть оно простое и надо вернуть `True`


In [ ]:
def is_prime0(n):
  if n <= 1: return False
  for d in range(2, n):
    if n % d == 0:
      return False
  return True

n = 36

2 * 18 = 36
3 * 12 = 36
4 * 9 = 36
6 * 6 = 36
9 * 4 = 36


In [ ]:
def is_prime(n):
  if n <= 1: return False
  for d in range(2, round(n**0.5) + 1):
    if n % d == 0:
      return False
  return True

In [ ]:
is_prime0(1)

False

In [ ]:
for n in range(1, 10000):
  if is_prime0(n) != is_prime(n):
    print(n)

## Задача 3. Чётное количество делителей

Напишите программу, которая ищет среди целых чисел, принадлежащих числовому отрезку $[126849; 126871]$, числа, имеющие ровно $4$ различных делителя. Выведите эти четыре делителя для каждого найденного числа в порядке возрастания.

Прикинем, получится ли обойтись тривиальным перебором всех возможных делителей для каждого числа из отрезка.

Всего чисел на отрезке $126871 - 126849 + 1 = 23$

In [ ]:
126871 - 126849 + 1

23

In [2]:
%%time
a, b = 126849, 126871
for n in range(a, b + 1):
    divisors = []
    for d in range(1, n + 1):
        if n % d == 0:
            divisors.append(d)
    if len(divisors) == 4:
        print(divisors)

[1, 3, 42283, 126849]
[1, 47, 2699, 126853]
[1, 5, 25373, 126865]
[1, 293, 433, 126869]
CPU times: user 331 ms, sys: 1.49 ms, total: 332 ms
Wall time: 332 ms


In [3]:
%%time
a, b = 126849, 126871
for n in range(a, b + 1):
    divisors = []
    for d in range(1, n + 1):
        if n % d == 0:
            divisors.append(d)
            if len(divisors) > 4: break
    if len(divisors) == 4:
        print(divisors)

[1, 3, 42283, 126849]
[1, 47, 2699, 126853]
[1, 5, 25373, 126865]
[1, 293, 433, 126869]
CPU times: user 118 ms, sys: 0 ns, total: 118 ms
Wall time: 121 ms


In [5]:
%%time
a, b = 126849, 226871
count = 0
for n in range(a, b + 1):
    if n % 10000 == 0: print(n)
    divisors = []
    for d in range(1, n + 1):
        if n % d == 0:
            divisors.append(d)
            if len(divisors) > 4: break
    if len(divisors) == 4:
        count += 1
print(count)

130000
140000
150000
160000
170000
180000
190000
200000
210000
220000
21648
CPU times: user 10min 45s, sys: 994 ms, total: 10min 46s
Wall time: 10min 47s


In [6]:
%%time
a, b = 126849, 226871
count = 0
for n in range(a, b + 1):
    if n % 10000 == 0: print(n)
    divisors = []
    for d1 in range(1, round(n**0.5) + 1):
        if n % d1 == 0:
            divisors.append(d1)
            d2 = n // d1
            if d1 < d2:
                divisors.append(d2)
            if len(divisors) > 4: break
    if len(divisors) == 4:
        count += 1
print(count)

130000
140000
150000
160000
170000
180000
190000
200000
210000
220000
21648
CPU times: user 1.79 s, sys: 7.02 ms, total: 1.79 s
Wall time: 1.8 s


## Задача 4. Нечётное количество делителей

Напишите программу, которая ищет среди целых чисел, принадлежащих числовому отрезку $[1820348; 2880927]$, числа, имеющие ровно $5$ различных делителей. Выведите эти делители для каждого найденного числа в порядке возрастания.

*Поляков, №31*, 

аналогичная задача в *Тренировочном варианте осень 2020, Статград*

### Наивное переборное решение: $O(n^2)$

Будем перебирать все числа $n$ из диапазона  $[a, b]$ и для каждого будем перебирать все возможные числа $d$ из диапазона $[1, n]$. Числа $d$, являющиеся делителями числа $n$, будем добавлять в список и считать.

> Можно обойтись перебором значений $d$ из диапазона $[2, n-1]$, так как $1$ и $n$ являются делителями любого числа $n$, но в этом случае надо не забыть скорректировать код с учётом присутствия двух дополнительных делителей.



In [ ]:
%%time
# a, b = 1820348, 2880927
a, b = 1820348, 1820357
results = []
for n in range(a, b + 1):
  divisors  = []
  for d in range(2, n):
    if n % d == 0:
      divisors.append(d)
  if len(divisors) == 3:
    results.append((n, divisors))
      
print(*results, sep='\n')


CPU times: user 2.39 s, sys: 4.66 ms, total: 2.4 s
Wall time: 2.42 s


### Идея оптимизации 1: от $O(n^2)$ к $O(n\sqrt{n})$

Все делители числа $n$ можно получить, перебирая только возможные значения меньшего делителя $d_1$, а парный к нему больший делитель получать как $d_2=\frac{n}{d_1}$. 

В том случае, если $d_1 = d_2 = d$ (при $n = d^2$), в список делителей надо добавить только один из них. 

In [ ]:
%%time

a, b = 1820348, 1920360
answers = []
for n in range(a, b + 1):
  divisors = []
  for d1 in range(2, int(n**0.5) + 1):
    if n % d1 == 0:
      divisors.append(d1)
      d2 = n // d1
      if d1 != d2:
        divisors.append(d2)
  
  if len(divisors) == 3:
    answers.append((n, sorted(divisors)))

print(*answers, sep='\n')

(1874161, [37, 1369, 50653])
CPU times: user 17.2 s, sys: 34.8 ms, total: 17.3 s
Wall time: 17.3 s


In [ ]:
%%time

a, b = 1820348, 1920360
answers = []
for n in range(a, b + 1):
  divisors = []
  for d1 in range(1, int(n**0.5) + 1):
    if n % d1 == 0:
      divisors.append(d1)
      d2 = n // d1
      if d1 != d2:
        divisors.append(d2)
      if len(divisors) > 5:
        break
  
  if len(divisors) == 5:
    answers.append(sorted(divisors))

print(*answers, sep='\n')

[1, 37, 1369, 50653, 1874161]
CPU times: user 5.14 s, sys: 7.55 ms, total: 5.14 s
Wall time: 5.17 s


In [ ]:
%%time

a, b = 1820348, 2880927
answers = []
for n in range(a, b + 1):
  if n % 100000 == 0: print(n)
  divisors = []
  for d1 in range(1, int(n**0.5) + 1):
    if n % d1 == 0:
      divisors.append(d1)
      d2 = n // d1
      if d1 != d2:
        divisors.append(d2)
      if len(divisors) > 5:
        break
  
  if len(divisors) == 5:
    answers.append(sorted(divisors))

print(*answers, sep='\n')

1900000
2000000
2100000
2200000
2300000
2400000
2500000
2600000
2700000
2800000
[1, 37, 1369, 50653, 1874161]
[1, 41, 1681, 68921, 2825761]
CPU times: user 59.7 s, sys: 111 ms, total: 59.9 s
Wall time: 1min


### Идея оптимизации 2: от $O(n\sqrt{n})$ к $O(n)$

Число $n$ может иметь нечетное чило делителей только в том случае, если он является квадратом некоторого целого числа.

То есть существует такое $d \in \Bbb{N}$ что $n = d^2$. В этом случае у числа $n$ делителями обязательно будут числа $\{1, d, d^2\}$, то есть как минимум три, а нетривиальными &mdash; как минимум один. 

> Напомним, что нетривиальным делителем числа $n$ называется делитель, отличный от $1$ и $n$

Если число $d$ простое, то кроме этих трех делителей других не будет. Если число $d$ составное, то любой нетривиальный делитель числа $d$

In [ ]:
%%time
a, b = 1820348, 2880927
a0, b0 = int(a**0.5), int(b**0.5) + 1
print(a0, b0)
answers = []
for d in range(a0, b0 + 1):
  n = d ** 2
  if n < a or n > b: continue
  
  divisors = []
  for d1 in range(1, int(n ** 0.5) + 1):
    if n % d1 == 0:
      divisors.append(d1)
      d2 = n // d1
      if d2 != d1:
        divisors.append(d2)
      if len(divisors) > 5: break
  
  if len(divisors) == 5:
    answers.append(sorted(divisors))
    
print(*answers, sep='\n')

1349 1698
[1, 37, 1369, 50653, 1874161]
[1, 41, 1681, 68921, 2825761]
CPU times: user 14.1 ms, sys: 1.99 ms, total: 16.1 ms
Wall time: 19.3 ms


In [ ]:
%%time
a, b = 1820348, 2880927
a0, b0 = int(a**0.25), int(b**0.25) + 1
answers = []
print(a0, b0)
for p in range(a0, b0 + 1):
  n = p ** 4
  if n < a or n > b: continue
  
  divisors = []
  for d1 in range(1, int(n ** 0.5) + 1):
    if n % d1 == 0:
      divisors.append(d1)
      d2 = n // d1
      if d2 != d1:
        divisors.append(d2)
      if len(divisors) > 5: break
  
  if len(divisors) == 5:
    answers.append(sorted(divisors))

print(*answers, sep='\n')

36 42
[1, 37, 1369, 50653, 1874161]
[1, 41, 1681, 68921, 2825761]
CPU times: user 2.1 ms, sys: 1e+03 ns, total: 2.1 ms
Wall time: 2.04 ms


In [ ]:
%%time
def is_prime(n):
    if n == 1: return False
    for d in range(2, int(n**0.5) + 1):
        if n % d == 0: return False
    return True

a, b = 1820348, 2880927
a0, b0 = int(a**0.25), int(b**0.25) + 1
answers = []
print(a0, b0)
for p in range(a0, b0 + 1):
  if not is_prime(p): continue

  n = p ** 4
  if n < a or n > b: continue
  answers.append([1, p, p**2, p**3, p**4])

print(*answers, sep='\n')

36 42
[1, 37, 1369, 50653, 1874161]
[1, 41, 1681, 68921, 2825761]
CPU times: user 515 µs, sys: 0 ns, total: 515 µs
Wall time: 326 µs


In [ ]:
%%time
def is_prime(n):
    if n == 1: return False
    for d in range(2, int(n ** 0.5) + 1):
        if n % d == 0: return False
    return True

a, b = 525784203, 728943762
a0, b0 = int(a ** 0.25), int(b ** 0.25) + 1
answers = []
for p in range(a0, b0 + 1):
    if not is_prime(p): continue

    n = p ** 4
    if a <= n <= b: 
        print(n, p**3)

607573201 3869893
705911761 4330747
CPU times: user 1.48 ms, sys: 0 ns, total: 1.48 ms
Wall time: 1.49 ms


In [ ]:
%%time
def is_prime(n):
  for d in range(2, int(n**0.5) + 1):
    if n % d == 0: return False
  return True

a, b = 100000000, 200000000
a0, b0 = int(a**0.25), int(b**0.25) + 1
answers = []
print(a0, b0)
for p in range(a0, b0 + 1):
  if not is_prime(p): continue

  n = p ** 4
  if n < a or n > b: continue
  answers.append([1, p, p**2, p**3, p**4])

print(*answers, sep='\n')

100 119
[1, 101, 10201, 1030301, 104060401]
[1, 103, 10609, 1092727, 112550881]
[1, 107, 11449, 1225043, 131079601]
[1, 109, 11881, 1295029, 141158161]
[1, 113, 12769, 1442897, 163047361]
CPU times: user 2.51 ms, sys: 0 ns, total: 2.51 ms
Wall time: 3.27 ms


In [ ]:
p = []
for i in range(1, 10):
  p.append(i**2)
print(p)

p = [i**2 for i in range(1, 10)]
print(p)

[1, 4, 9, 16, 25, 36, 49, 64, 81]
[1, 4, 9, 16, 25, 36, 49, 64, 81]


In [ ]:
p = []
for i in range(1, 10):
  if i != 5:
    p.append(i**2)
print(p)

p = [i**2 for i in range(1, 10) if i != 5]
print(p)

[1, 4, 9, 16, 36, 49, 64, 81]
[1, 4, 9, 16, 36, 49, 64, 81]


In [ ]:
%%time
a, b = 1820348, 2880927
a0, b0 = int(a**0.25), int(b**0.25) + 1
ans = []
print(a0, b0)
for k in range(a0, b0 + 1):
  i = k**4
  if i < a or i > b: continue
  count = 0
  ansBeta = []
  for j in range(2, math.ceil(math.sqrt(k)) + 1):
    if (i % j == 0):
      count +=1
      break
  
  if (count == 0):
    print([k**n for n in range(5)])
    

36 42
[1, 37, 1369, 50653, 1874161]
[1, 41, 1681, 68921, 2825761]
CPU times: user 1.46 ms, sys: 0 ns, total: 1.46 ms
Wall time: 1.47 ms


Как напечатать список чисел в поэлемнтно в одну строку через пробел

In [ ]:
for n in range(5):
  print(k**n, end=' ')

1 42 1764 74088 3111696 

In [ ]:
print(*[k**n for n in range(5)])

1 42 1764 74088 3111696


In [ ]:
 p = [k**n for n in range(5)]
print(p[0], p[1], p[2], p[3], p[4])

1 42 1764 74088 3111696


## Задача 5. Заданное количество чётных делителей

Найдите все натуральные числа, принадлежащие отрезку
$[101\ 000\ 000; 102\ 000\ 000]$, у которых ровно три различных чётных делителя.
В ответе перечислите найденные числа в порядке возрастания.

*Тренировочная работа №3 по ИНФОРМАТИКЕ, 11 класс, 2 февраля 2021 года, Статград*

### Наивное переборное решение:  $O(n^2)$

В этой задаче рассчитывать на это решение не приходится, так как для чисел порядка $10^8$ и количества чисел порядка $10^6$ количество операций будет порядка $10^{14}$.

Тем не менее, попробуем.

Сразу учтём, что:
- число $n$ с чётными делителями может быть только чётным, поэтому в цикле по $n$ будем перебирать только чётные числа: `for n in range(a, b + 1, 2)`
- в цикле перебора делителей тоже будем перебирать только чётные числа: `for d in range(2, n + 1, 2):`
- если будет найдено более трёх делителей, то остальные делители для данного $n$ можно не находить
- чтобы можно было дождаться окончания работы программы возьмём сначала уменьшенный диапазон чисел

In [ ]:
%%time
a = 101000000
b = 101000010

for n in range(a, b + 1, 2):
  delimeters = set()
  for d in range(2, n + 1, 2):
    if n % d == 0:
      delimeters.add(d)
      if len(delimeters) > 3:
        break
  if len(delimeters) == 3:
    print(n, delimeters)

CPU times: user 13.2 s, sys: 21 ms, total: 13.2 s
Wall time: 13.3 s


### От $O(n^2)$ к $O(n\sqrt{n})$

Но не всегда можно сразу догадаться даже до того, что достаточно перебирать только чётные числа $n$.

Зато оптимизация перебора делителей не до $n$, а только до $\sqrt{n}$ более известная. Воспользуемся ей.

In [ ]:
%%time
from math import ceil
a = 101000000
b = 101100000

for n in range(a, b + 1, 1):
  delimeters = set()
  for d1 in range(1, round(n**0.5) + 1):
    if n % d1 == 0:
      d2 = n // d1
      if d1 % 2 == 0: delimeters.add(d1)
      if d2 % 2 == 0: delimeters.add(d2)

      if len(delimeters) > 3:
        break
  if len(delimeters) == 3:
    print(n, delimeters)

101075762 {101075762, 2, 14218}
CPU times: user 1min 8s, sys: 82 ms, total: 1min 8s
Wall time: 1min 8s


In [ ]:
%%time
from math import ceil
a = 101000000
b = 102000000

for i in range(a, b + 1, 2):
    if i % 4 == 0:
        continue
    x = []
    for n in range(1, int(i ** 0.5) + 1):
        if i % n == 0:
            if n % 2 == 0:
                x.append(n)
            d = i // n
            if d % 2 == 0 and d != n:
                x.append(d)
            if len(x) > 3:
                break
    if len(x) == 3:
        print(i, x)

101075762 [101075762, 2, 14218]
101417282 [101417282, 2, 14242]
101588258 [101588258, 2, 14254]
101645282 [101645282, 2, 14258]
CPU times: user 48.9 s, sys: 58.9 ms, total: 49 s
Wall time: 49.2 s


In [ ]:
101075762 == 14218**2 // 2 

True

Число имеет три чётных делителя тогда и только тогда, когда оно представимо в виде 
$$n = 2 \cdot p \cdot p = 2p^2$$
где $p$ &mdash; простое число. Три чётных делителя будут числа $[2, 2 \cdot p, n]$. Найдем диапазон, которому могут принадлежать такие числа $p$, чтобы числа $n=2p^2$ принадлежали диапазону $[a, b]$:

$$a \le n \le b \\
a \le 2p^2 \le b \\
\sqrt{\frac{a}{2}} \le p \le \sqrt{\frac{b}{2}} \\
$$

In [ ]:
%%time
a, b = 101000000, 102000000

a0, b0 = int((a // 2)**0.5), int(((b + 1) // 2)**0.5) + 1

answers = []

for p in range(a0, b0 + 1):
    if not is_prime(p): continue
    
    n = p * p * 2
    if a <= n <= b:
        answers.append(n)

print(sorted(answers))

[101075762, 101417282, 101588258, 101645282]
CPU times: user 389 µs, sys: 0 ns, total: 389 µs
Wall time: 323 µs


## Задача 6. Заданное количество нечётных делителей

Найдите все натуральные числа, принадлежащие отрезку $[35\ 000\ 000; 40\ 000\ 000]$,
у которых ровно пять различных нечётных делителей (количество чётных
делителей может быть любым). В ответе перечислите найденные числа
в порядке возрастания.

In [ ]:
%%time
answers = []

for n in range(35000000, 40000000):
  counter = 0
  if n % (100000) == 0:
    print(n)
  for d1 in range(1, int(n**0.5) + 1):
    if n % d1 == 0:
       if d1 % 2 != 0:
          counter += 1
       d2 = n // d1
       if d2 != d1 and d2 % 2 != 0:
          counter += 1
       if counter > 5:
          break
  if counter == 5:
    answers.append(i)

print(answers)

35000000
35100000
35200000
35300000
35400000
35500000
35600000
35700000
35800000
35819648
35900000
36000000
36100000
36200000
36300000
36400000
36500000
36600000
36700000
36800000
36900000
37000000
37100000
37200000
37300000
37400000
37500000
37600000
37700000
37800000
37900000
38000000
38100000
38200000
38300000
38400000
38500000
38600000
38700000
38800000
38900000
38950081
39000000
39037448
39100000
39200000
39300000
39337984
39400000
39500000
39600000
39700000
39800000
39900000
CPU times: user 28min 19s, sys: 1.76 s, total: 28min 20s
Wall time: 28min 23s


In [ ]:
%%time
answers = []

for i in range(35000000, 40000000):
  counter = 0
  if i % 500000 == 0:
    print(i)

  n = i
  while n % 2 == 0:
    n //= 2

  for d1 in range(1, round(n**0.5) + 1):
    if n % d1 == 0:
       counter += 1
       d2 = n // d1
       if d2 != d1:
         counter += 1
       if counter > 5:
          break
  if counter == 5:
    answers.append(i)

print(answers)

35000000
35500000
36000000
36500000
37000000
37500000
38000000
38500000
39000000
39500000
[35819648, 38950081, 39037448, 39337984]
CPU times: user 21min 24s, sys: 1.86 s, total: 21min 26s
Wall time: 21min 30s


In [ ]:
import threading

In [ ]:
%%time
def writer(a, b, answers):
    for i in range(a, b + 1):
        counter = 0
        if i % 500000 == 0:
            print(i)

        n = i
        while n % 2 == 0:
            n //= 2

        for d1 in range(1, round(n**0.5) + 1):
            if n % d1 == 0:
                counter += 1
                d2 = n // d1
                if d2 != d1:
                    counter += 1
                if counter > 5:
                    break
        if counter == 5:
            answers.append(i)

t1 = threading.Thread(target=writer, args=(35000000, 37500000, answers))
t2 = threading.Thread(target=writer, args=(37500001, 40000000, answers))

t1.start()
t2.start()

t1.join()
t2.join()

print(answers)  

35000000
35500000
38000000
36000000
38500000
36500000
39000000
37000000
39500000
37500000
40000000
[39337984, 35819648, 39037448, 38950081, 35819648, 38950081, 39037448, 39337984]
CPU times: user 15min 5s, sys: 6 s, total: 15min 11s
Wall time: 15min 13s


In [ ]:
%%time
def writer(a, b, answers):
    for i in range(a, b + 1):
        counter = 0
        if i % 500000 == 0:
            print(i)

        n = i
        while n % 2 == 0:
            n //= 2

        for d1 in range(1, round(n**0.5) + 1):
            if n % d1 == 0:
                counter += 1
                d2 = n // d1
                if d2 != d1:
                    counter += 1
                if counter > 5:
                    break
        if counter == 5:
            answers.append(i)

# init threads
answers = []
threads = []
a, b = 35000000, 40000000
n_threads = 10
h = (b - a + n_threads - 1) // n_threads 

for t in range(n_threads):
    a0 = a + h * t
    b0 = a + h * (t + 1)
    print(f'Thread[{t}] work with {(a0, b0)}')
    threads.append(threading.Thread(target=writer, args=(a0, b0, answers)))


# start threads
for t in range(n_threads):
    threads[t].start()

# join threads to the main thread
for t in range(n_threads):
    threads[t].join()

print(answers)  

Thread[0] work with (35000000, 35500000)
Thread[1] work with (35500000, 36000000)
Thread[2] work with (36000000, 36500000)
Thread[3] work with (36500000, 37000000)
Thread[4] work with (37000000, 37500000)
Thread[5] work with (37500000, 38000000)
Thread[6] work with (38000000, 38500000)
Thread[7] work with (38500000, 39000000)
Thread[8] work with (39000000, 39500000)
Thread[9] work with (39500000, 40000000)
35000000
35500000
36000000
36500000
37000000
37500000
38000000
38500000
39000000
39500000
35500000
36500000
37000000
38000000
39500000
36000000
39000000
37500000
38500000
40000000
[39037448, 35819648, 39337984, 38950081]
Wall time: 12min 6s


```cpp
#include <iostream>
#include <vector>

int main() {
    std::vector<int> answers;

    for (int i = 35000000; i <= 40000000; i++) {
        int counter = 0;
        if (i % 500000 == 0) 
            std::cout << i << std::endl;

        int n = i;
        while (n % 2 == 0)
            n /= 2;

        for (int d1 = 1; d1*d1 <= n; d1++) {
            if (n % d1 == 0) {
                if (d1 % 2 == 1) { counter += 1; }
                int d2 = n / d1;
                if (d2 != d1 && d2 % 2 == 1) { counter += 1; }
                if (counter > 5) break;
            }
        }
        if (counter == 5)
            answers.push_back(i);
    }
    for (auto n : answers) { std::cout << n << ' '; }
    return 0;
}
```

Число имеет пять различных нечётных делителей тогда и только тогда, когда оно представимо в виде 
$$n = 2^k \cdot p \cdot p \cdot p \cdot p = 2^kp^4$$
где $p$ &mdash; простое число, $k$ &mdash; целое число, $k \ge 0$. Пять нечётных делителей будут числа $[1,\ p,\ p^2,\ p^3,\ p^4]$. Найдем диапазон, которому могут принадлежать такие числа $p$, чтобы числа $n=2^kp^4$ принадлежали диапазону $[a, b]$.

Поскольку $k$ может быть сколь угодно большим, то минимальное $p$ можно взять равным наименьшему нечётному простому числу, то есть $3$. Максимальное $p$ будет соответствовать наименьшему $k = 0$:

$$ n \le b \\
2^0p^4 \le b \\
p \le \sqrt[4]{b} \\
$$

Тогда остается перебрать простые числа $p$ из диапазона $[3, \sqrt[4]{b}]$ и для каждого подобрать такие значения $k$, чтобы выполнялось неравенство
$$a \le 2^kp^4 \le b$$

In [ ]:
%%time
def is_prime(n):
  '''Проверка числа `n` на простоту'''
  if n == 1: return False
  for d in range(2, int(n**0.5) + 1):
    if n % d == 0: return False
  return True

a, b = 35000000, 40000000
a0, b0 = 3, int(b**0.25) + 1
print(a0, b0)

answers = []
for p in range(a0, b0 + 1):
  if not is_prime(p): continue
  
  n = p**4  # Начинаем с k=0

  while True:
    if a <= n <= b:
      answers.append(n)
    n *= 2  # Берем следующее k
    if n > b: break   

print(*sorted(answers))

3 80
35819648 38950081 39037448 39337984
CPU times: user 1.14 ms, sys: 0 ns, total: 1.14 ms
Wall time: 930 µs


In [ ]:
%%time
for n in answer:
  divisors = []
  for d in range(1, n+1):
    if n % d == 0:
      divisors.append(d)
  print(n, divisors)
  print(n, [d for d in divisors if d % 2 == 1])

39337984 [1, 2, 4, 7, 8, 14, 16, 28, 32, 49, 56, 64, 98, 112, 128, 196, 224, 256, 343, 392, 448, 512, 686, 784, 896, 1024, 1372, 1568, 1792, 2048, 2401, 2744, 3136, 3584, 4096, 4802, 5488, 6272, 7168, 8192, 9604, 10976, 12544, 14336, 16384, 19208, 21952, 25088, 28672, 38416, 43904, 50176, 57344, 76832, 87808, 100352, 114688, 153664, 175616, 200704, 307328, 351232, 401408, 614656, 702464, 802816, 1229312, 1404928, 2458624, 2809856, 4917248, 5619712, 9834496, 19668992, 39337984]
39337984 [1, 7, 49, 343, 2401]
35819648 [1, 2, 4, 8, 16, 23, 32, 46, 64, 92, 128, 184, 368, 529, 736, 1058, 1472, 2116, 2944, 4232, 8464, 12167, 16928, 24334, 33856, 48668, 67712, 97336, 194672, 279841, 389344, 559682, 778688, 1119364, 1557376, 2238728, 4477456, 8954912, 17909824, 35819648]
35819648 [1, 23, 529, 12167, 279841]
39037448 [1, 2, 4, 8, 47, 94, 188, 376, 2209, 4418, 8836, 17672, 103823, 207646, 415292, 830584, 4879681, 9759362, 19518724, 39037448]
39037448 [1, 47, 2209, 103823, 4879681]
38950081 [1, 7

In [ ]:
k = 11
while 2**k * 3**4 < 165000000:
  k += 1
  if 2**k * 3**4 > 175000000:
    break
k

21

In [ ]:
2**21 * 3**4, (2**21 * 3**4)**0.5

(169869312, 13033.392190830444)

In [ ]:
%%time
a, b = 35000000, 40000000
a0, b0 = 3, int(b**0.25) + 1
print(a0, b0)

answers = []
for p in range(a0, b0 + 1):
    n = p**4
    k = 0
    while n <= b:
        if n >= a:
            divisors = []
            for d1 in range(1, round(n**0.5)+1):
                if n % d1 == 0:
                    d2 = n // d1
                    if d1 % 2 == 1:
                        divisors.append(d1)
                    if d2 % 2 == 1 and d2 != d1:
                        divisors.append(d2)
                    if len(divisors) > 5:
                        break
            if len(divisors) == 5:
                answers.append(n) 
        k += 1
        n *= 2

print(sorted(set(answers)))       

3 80
[35819648, 38950081, 39037448, 39337984]
CPU times: user 7.41 ms, sys: 0 ns, total: 7.41 ms
Wall time: 7.34 ms


## Задача 7. Поиск пар делителей с заданной разностью

Рассмотрим произвольное натуральное число, представим его всеми возможными способами в виде произведения двух натуральных чисел и найдём для каждого такого произведения разность сомножителей. Например, для числа $16$ получим: $16 = 16 \cdot 1 = 8 \cdot 2 = 4 \cdot 4$, множество разностей содержит числа $15$, $6$ и $0$. Найдите все натуральные числа, принадлежащие отрезку $[1\ 000\ 000; 2\ 000\ 000]$, у которых составленное описанным способом множество разностей будет содержать не меньше трёх элементов, не превышающих $100$. В ответе перечислите найденные числа в порядке возрастания.

In [ ]:
%%time
a, b = 1000000, 2000000
answers = []

for n in range(a, b + 1):
  if n % 100000 == 0: print(n)
  count = 0
  for d1 in range(1, round(n**0.5) + 1):
    if n % d1 == 0:
      d2 = n // d1
      diff = d2 - d1
      if d2 < d1: break
      if diff <= 100:
        count += 1
        if count > 3: break
  if count >= 3:
    answers.append(n)

print(answers)  

1000000
1100000
1200000
1300000
1400000
1500000
1600000
1700000
1800000
1900000
2000000
[1113840, 1179360, 1208844, 1499400]
CPU times: user 2min 28s, sys: 212 ms, total: 2min 28s
Wall time: 2min 28s


In [ ]:
%%time
a, b = 1000000, 1000020
answers = []

for n in range(a, b + 1):
  print(n)
  count = 0
  for d1 in range(1, round(n**0.5) + 1):
    if n % d1 == 0:
      d2 = n // d1
      diff = d2 - d1
      print(d1, d2, diff)
      if d2 < d1: break
      if diff <= 100:
        count += 1
  if count >= 3:
    answers.append(n)

print(answers) 

1000000
1 1000000 999999
2 500000 499998
4 250000 249996
5 200000 199995
8 125000 124992
10 100000 99990
16 62500 62484
20 50000 49980
25 40000 39975
32 31250 31218
40 25000 24960
50 20000 19950
64 15625 15561
80 12500 12420
100 10000 9900
125 8000 7875
160 6250 6090
200 5000 4800
250 4000 3750
320 3125 2805
400 2500 2100
500 2000 1500
625 1600 975
800 1250 450
1000 1000 0
1000001
1 1000001 1000000
101 9901 9800
1000002
1 1000002 1000001
2 500001 499999
3 333334 333331
6 166667 166661
1000003
1 1000003 1000002
1000004
1 1000004 1000003
2 500002 500000
4 250001 249997
53 18868 18815
89 11236 11147
106 9434 9328
178 5618 5440
212 4717 4505
356 2809 2453
1000005
1 1000005 1000004
3 333335 333332
5 200001 199996
15 66667 66652
163 6135 5972
409 2445 2036
489 2045 1556
815 1227 412
1000006
1 1000006 1000005
2 500003 500001
7 142858 142851
14 71429 71415
1000007
1 1000007 1000006
29 34483 34454
1000008
1 1000008 1000007
2 500004 500002
3 333336 333333
4 250002 249998
6 166668 166662
8 125001

In [ ]:
%%time
a, b = 1000000, 2000000
answers = []

for n in range(a, b + 1):
  if n % 100000 == 0: print(n)
  count = 0
  for d1 in range(round(n**0.5), 0, -1):
    if n % d1 == 0:
      d2 = n // d1
      diff = d2 - d1
      if diff <= 100:
        count += 1
      else:
        break
  if count >= 3:
    answers.append(n)

print(answers) 

1000000
1100000
1200000
1300000
1400000
1500000
1600000
1700000
1800000
1900000
2000000
[1113840, 1179360, 1208844, 1499400]
CPU times: user 1min 43s, sys: 164 ms, total: 1min 43s
Wall time: 1min 43s


In [ ]:
n = 2053440
divisors = []
for d1 in range(round(n**0.5), 0, -1):
  if n % d1 == 0:
    d2 = n // d1
    divisors.append((d1, d2))

print(*divisors, sep='\n')

1414.213562373095
(1426, 1440)
(1395, 1472)
(1380, 1488)
(1240, 1656)
(1116, 1840)
(1104, 1860)
(1035, 1984)
(992, 2070)
(960, 2139)
(930, 2208)
(920, 2232)
(828, 2480)
(744, 2760)
(736, 2790)
(720, 2852)
(713, 2880)
(690, 2976)
(620, 3312)
(576, 3565)
(558, 3680)
(552, 3720)
(496, 4140)
(480, 4278)
(465, 4416)
(460, 4464)
(414, 4960)
(372, 5520)
(368, 5580)
(360, 5704)
(345, 5952)
(320, 6417)
(310, 6624)
(288, 7130)
(279, 7360)
(276, 7440)
(248, 8280)
(240, 8556)
(230, 8928)
(207, 9920)
(192, 10695)
(186, 11040)
(184, 11160)
(180, 11408)
(160, 12834)
(155, 13248)
(144, 14260)
(138, 14880)
(124, 16560)
(120, 17112)
(115, 17856)
(96, 21390)
(93, 22080)
(92, 22320)
(90, 22816)
(80, 25668)
(72, 28520)
(69, 29760)
(64, 32085)
(62, 33120)
(60, 34224)
(48, 42780)
(46, 44640)
(45, 45632)
(40, 51336)
(36, 57040)
(32, 64170)
(31, 66240)
(30, 68448)
(24, 85560)
(23, 89280)
(20, 102672)
(18, 114080)
(16, 128340)
(15, 136896)
(12, 171120)
(10, 205344)
(9, 228160)
(8, 256680)
(6, 342240)
(5, 410688

In [ ]:
# from collections import Counter
counts = dict()

for d1 in range(10, 21):
  for d2 in range(d1, 21):
    n = d1 * d2
    diff = d2 - d1
    if diff > 10:
      break
    if n in counts:
      counts[n] = counts[n] + 1
    else:
      counts[n] = 1

for n in counts:
  if counts[n] >= 2:
    print(n, counts[n])

180 2
240 2


In [ ]:
%%time
from collections import Counter
counts = Counter()

a, b = 1000000, 2000000

for d1 in range(round(a**0.5)-100, round(b**0.5) + 1):
  for d2 in range(d1, round(b**0.5) + 100 + 1):
    n = d1 * d2
    if n < a: continue
    if n > b: break
    diff = d2 - d1
    if diff > 100:
      break
    counts[n] += 1

for n in counts:
  if counts[n] >= 3:
    print(n, counts[n])

1113840 3
1179360 3
1208844 3
1499400 3
CPU times: user 47.2 ms, sys: 997 µs, total: 48.2 ms
Wall time: 48.5 ms


In [ ]:
%%time
from collections import Counter
counts = Counter()

a, b = 1000000, 2000000

for d1 in range(round(a**0.5)-100, round(b**0.5) + 1):
  for d2 in range(d1, d1 + 100 + 1):
    n = d1 * d2
    if n > b: break
    diff = d2 - d1
    counts[n] += 1

for n in counts:
  if counts[n] >= 3:
    print(n, counts[n])

1113840 3
1179360 3
1208844 3
1499400 3
CPU times: user 42.1 ms, sys: 994 µs, total: 43 ms
Wall time: 47.7 ms


In [ ]:
1113840, 1179360, 1208844, 1499400

In [ ]:
2000000**0.5

1414.213562373095

In [ ]:
round(7000000**0.5)

2646

In [ ]:
round(7000000**0.5)*round(7000000**0.5)

7001316

In [ ]:
%%time
a, b = 2000000, 3000000
answers = []

for n in range(a, b + 1):
  if n % 100000 == 0: print(n)
  count = 0
  for d1 in range(round(n**0.5), 0, -1):
    if n % d1 == 0:
      d2 = n // d1
      diff = d2 - d1
      if diff <= 115:
        count += 1
      else:
        break
  if count >= 3:
    answers.append(n)

print(answers) 

2000000
2100000
2200000
2300000
2400000
2500000
2600000
2700000
2800000
2900000
3000000
[2053440, 2098080, 2328480, 2638944]
CPU times: user 2min 25s, sys: 206 ms, total: 2min 25s
Wall time: 2min 26s


In [ ]:
%%time
from collections import Counter
counts = Counter()

a, b = 2000000, 3000000

for d1 in range(round(a**0.5) - 115, round(b**0.5) + 1):
  for d2 in range(d1, d1 + 115 + 1):
    n = d1 * d2
    if n < a: continue
    if n > b: break
    counts[n] += 1

for n in counts:
  if counts[n] >= 3:
    print(n, counts[n])

2053440 3
2098080 3
2328480 3
2638944 3
CPU times: user 45.1 ms, sys: 1.01 ms, total: 46.1 ms
Wall time: 47 ms


139)	Рассмотрим произвольное натуральное число, представим его всеми возможными способами в виде произведения двух натуральных чисел и найдём для каждого такого произведения разность сомножителей. Например, для числа 18 получим: 18 = 18*1 = 9*2 = 6*3, множество разностей содержит числа 17, 7 и 3. Подходящей будем называть пару сомножителей, разность между которыми не превышает 120. Найдите все натуральные числа, принадлежащие отрезку [2000000; 3000000], у которых есть не менее трёх подходящих пар сомножителей. В ответе перечислите найденные числа в порядке возрастания, справа от каждого запишите наибольший из всех сомножителей, образующих подходящие пары.

In [ ]:
%%time
a, b = 2000000, 3000000

for n in range(a, b + 1):
    k, md = 0, 0
    for d in range(int(n**0.5) - 60, int(n**0.5) + 1):
        if n % d == 0:
            k += 1
            md = max(md, n // d)
    if k > 2:
        print(n, md)

2053440 1488
2098080 1504
2328480 1584
2620800 1680
2638944 1683
2692800 1700
CPU times: user 9.76 s, sys: 23.5 ms, total: 9.78 s
Wall time: 9.85 s


In [ ]:
%%time
a, b = 2000000, 3000000

for n in range(a, b + 1):
    k, md = 0, 0
    for d in range(int(n**0.5) - 120, int(n**0.5) + 1):
        if n % d == 0 and n // d - d <= 120:
            k += 1
            md = max(md, n // d)
    if k > 2:
        print(n, md)

2053440 1488
2098080 1504
2328480 1584
2620800 1680
2638944 1683
2692800 1700
CPU times: user 17.6 s, sys: 26.6 ms, total: 17.7 s
Wall time: 17.8 s


In [ ]:
a**0.5, b**0.5, int(a**0.5 - b**0.5) + 1

(1414.213562373095, 1732.0508075688772, -316)

In [ ]:
for d1 in range(7, 13):
    for d2 in range(d1, d1 + 11):
        print(d1, '*', d2, '=', d1 * d2, '| diff =', d2 - d1, '| sqrt(n) - d1 =', '{:.2f}'.format((d1*d2)**0.5 - d1))

7 * 7 = 49 | diff = 0 | sqrt(n) - d1 = 0.00
7 * 8 = 56 | diff = 1 | sqrt(n) - d1 = 0.48
7 * 9 = 63 | diff = 2 | sqrt(n) - d1 = 0.94
7 * 10 = 70 | diff = 3 | sqrt(n) - d1 = 1.37
7 * 11 = 77 | diff = 4 | sqrt(n) - d1 = 1.77
7 * 12 = 84 | diff = 5 | sqrt(n) - d1 = 2.17
7 * 13 = 91 | diff = 6 | sqrt(n) - d1 = 2.54
7 * 14 = 98 | diff = 7 | sqrt(n) - d1 = 2.90
7 * 15 = 105 | diff = 8 | sqrt(n) - d1 = 3.25
7 * 16 = 112 | diff = 9 | sqrt(n) - d1 = 3.58
7 * 17 = 119 | diff = 10 | sqrt(n) - d1 = 3.91
8 * 8 = 64 | diff = 0 | sqrt(n) - d1 = 0.00
8 * 9 = 72 | diff = 1 | sqrt(n) - d1 = 0.49
8 * 10 = 80 | diff = 2 | sqrt(n) - d1 = 0.94
8 * 11 = 88 | diff = 3 | sqrt(n) - d1 = 1.38
8 * 12 = 96 | diff = 4 | sqrt(n) - d1 = 1.80
8 * 13 = 104 | diff = 5 | sqrt(n) - d1 = 2.20
8 * 14 = 112 | diff = 6 | sqrt(n) - d1 = 2.58
8 * 15 = 120 | diff = 7 | sqrt(n) - d1 = 2.95
8 * 16 = 128 | diff = 8 | sqrt(n) - d1 = 3.31
8 * 17 = 136 | diff = 9 | sqrt(n) - d1 = 3.66
8 * 18 = 144 | diff = 10 | sqrt(n) - d1 = 4.00
9 * 

In [ ]:
from collections import Counter
counts = Counter()

for d1 in range(7, 13):
    for d2 in range(d1, d1 + 11):
        n = d1 * d2
        # print(d1, '*', d2, '=', d1 * d2, '| diff =', d2 - d1)
        counts[n] += 1

print(counts)

Counter({144: 3, 112: 2, 120: 2, 180: 2, 49: 1, 56: 1, 63: 1, 70: 1, 77: 1, 84: 1, 91: 1, 98: 1, 105: 1, 119: 1, 64: 1, 72: 1, 80: 1, 88: 1, 96: 1, 104: 1, 128: 1, 136: 1, 81: 1, 90: 1, 99: 1, 108: 1, 117: 1, 126: 1, 135: 1, 153: 1, 162: 1, 171: 1, 100: 1, 110: 1, 130: 1, 140: 1, 150: 1, 160: 1, 170: 1, 190: 1, 200: 1, 121: 1, 132: 1, 143: 1, 154: 1, 165: 1, 176: 1, 187: 1, 198: 1, 209: 1, 220: 1, 231: 1, 156: 1, 168: 1, 192: 1, 204: 1, 216: 1, 228: 1, 240: 1, 252: 1, 264: 1})


In [ ]:
from collections import Counter
counts = Counter()

for d1 in range(7, 13):
    for d2 in range(d1, d1 + 11):
        n = d1 * d2
        # print(d1, '*', d2, '=', d1 * d2, '| diff =', d2 - d1)
        counts[n] += 1

for n in counts:
    if counts[n] > 2:
        print(n)

144


In [ ]:
from collections import Counter
counts = Counter()
max_divisors = Counter()

for d1 in range(7, 13):
    for d2 in range(d1, d1 + 11):
        n = d1 * d2
        # print(d1, '*', d2, '=', d1 * d2, '| diff =', d2 - d1)
        counts[n] += 1
        if d2 > max_divisors[n]:
            max_divisors[n] = d2

for n in counts:
    if counts[n] > 2:
        print(n, max_divisors[n])

144 18


In [ ]:
%%time
a, b = 49, 264

for n in range(a, b + 1):
    k, md = 0, 0
    for d in range(int(n**0.5) - 5, int(n**0.5)):
        if n % d == 0 and n // d - d <= 10:
            k += 1
            md = max(md, n // d)
    if k > 2:
        print(n, md)

CPU times: user 437 µs, sys: 0 ns, total: 437 µs
Wall time: 440 µs


In [ ]:
%%time
a, b = 49, 264

for n in range(a, b + 1):
    k, md = 0, 0
    for d in range(int(n**0.5) - 5, int(n**0.5) + 1):
        if n % d == 0 and n // d - d <= 10:
            k += 1
            md = max(md, n // d)
    if k > 2:
        print(n, md)

144 18
CPU times: user 2.43 ms, sys: 0 ns, total: 2.43 ms
Wall time: 2.63 ms


In [ ]:
from collections import Counter
counts = Counter()
max_divisors = Counter()

a, b = 49, 264
diff = 10

for d1 in range(int(a**0.5) - diff, int(b**0.5) + 1):
    for d2 in range(d1, d1 + diff + 1):
        n = d1 * d2
        if n < a or n > b: continue
        # print(d1, '*', d2, '=', d1 * d2, '| diff =', d2 - d1)
        counts[n] += 1
        if d2 > max_divisors[n]:
            max_divisors[n] = d2

for n in counts:
    if counts[n] > 2:
        print(n, max_divisors[n])

144 18


In [ ]:
%%time
from collections import Counter
counts = Counter()
max_divisors = Counter()

a, b = 2000000, 3000000
diff = 120

for d1 in range(int(a**0.5) - diff, int(b**0.5) + 1):
    for d2 in range(d1, d1 + diff + 1):
        n = d1 * d2
        if n < a or n > b: continue
        # print(d1, '*', d2, '=', d1 * d2, '| diff =', d2 - d1)
        counts[n] += 1
        if d2 > max_divisors[n]:
            max_divisors[n] = d2

for n in counts:
    if counts[n] > 2:
        print(n, max_divisors[n])

2053440 1488
2098080 1504
2328480 1584
2620800 1680
2638944 1683
2692800 1700
CPU times: user 66.7 ms, sys: 1.97 ms, total: 68.7 ms
Wall time: 67.7 ms


Два числа называются дружественными если сумма собственных делителей (то есть всех положительных делителей, отличных от самого́ числа) любого их них равна другому числу. Например, числа 220 и 284 дружественные.

Выведите в порядке возрастания числа в диапазоне [2; 30000], имеющие дружественное число, большее чем само это число, и через пробел это дружественное число. Каждое следующие число из указанного диапазона выводите на новой строке.

In [ ]:
%%time
numbers = dict()

a, b = 2, 30000

for n in range(a, b + 1):
    divisors = set()
    for d1 in range(2, int(n ** 0.5) + 1):
        if n % d1 == 0:
            divisors.add(d1)
            divisors.add(n // d1)
        
    s = sum(divisors) + 1
    numbers[s] = numbers.get(s, []) + [n]

print(numbers)

for s in numbers:
    for t in numbers[s]:
        if t > s and t in numbers and s in numbers[t]:
            print(s, t)


{1: [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 1217

## Теория

Представим число $n$ в виде произведения простых сомножителей:
$$n = p_1^{\alpha_1} \cdot p_2^{\alpha_1} \dots p_k^{\alpha_k}$$

где $p_1, p_2, \dots p_k$ &mdash; простые числа, $\alpha_1, \alpha_2, \dots \alpha_k$ &mdash; натуральные числа.

Такое представление единственно для любого натурального числа.

Например, $175760 = 2^2 \cdot 5^1 \cdot 13^3$, и другим способом представить его в виде произведения проcтых сомножителей нельзя.

Рассмотрим некоторые частные случаи такого представления.

### Случай $k = 1, \alpha_1 = 1$

В этом случае $n = p_1$, то есть простое число. У такого числа ровно два делителя: $\{1,\ p_1\}$. Например, $n = 17$ имеет ровно **два** делителя $\{1, 17\}$

### Случай $k = 1, \alpha_1 > 1$

В этом случае $n = p_1^{\alpha_1}$, то есть простое число в некоторой степени. У такого числа ровно ${\alpha_1+1}$ делителей &mdash; все степени числа $p_1$ от $p_1^0 = 1$ до $p^{\alpha_1}$: $\{1,\ p_1^1,\ p_1^2,\ \dots \ p_1^{\alpha_1}\}$.

Если $\alpha_1$ *чётное* число, то тогда количество делителей числа $n$ будет **нечётным**.

### Случай $k = 2, \alpha_1 = 1, \alpha_2 = 1$

В этом случае $n = p_1 \cdot p_2$, то есть произведение двух простых чисел. У такого числа будет ровно $4$ делителя: $\{1,\ p_1,\ p_2,\ p_1 p_2\}$. Других случаев, при которых у числа $n$
 **четыре** делителя не будет.

 ### Случай $k = 2, \alpha_1 \ge 1, \alpha_2 \ge 1$

 В этом случае $n = p_1^{\alpha_1} \cdot p_2^{\alpha_2}$. У такого числа количество делителей можно вычислить следующим образом: если мы возьмём число $p_1$ любой степени от $0$ до $\alpha_1$ и умножим его на $p_2$ любой степени от $0$ до $\alpha_2$, то результат будет являться делителем числа $n$.

Например, для числа $12 = 2^2 \cdot 3^1$ возьмем множество значений $\beta_1 = \{0,1,2\}$ и $\beta_2 = \{0,1\}$. Составим все возможные комбинации из элементов этих двух множеств. Таких комбинаций будет $3 \cdot 2 = 6$ и получим все шесть делителей числа $12$:
$$\begin{array}{c|c|c}
№ & \beta_1 & \beta_2 & p_1^{\beta_1} & p_2^{\beta_2} & делитель & вид\\
\hline
1 & 0 & 0 & 1 & 1 & 1 & 2^0 \cdot 3^0 \\
2 & 1 & 0 & 2 & 1 & 2 & 2^1 \cdot 3^0 \\
3 & 2 & 0 & 4 & 1 & 4 & 2^2 \cdot 3^0 \\
4 & 0 & 1 & 1 & 3 & 3 & 2^2 \cdot 3^1 \\
5 & 1 & 1 & 2 & 3 & 6 & 2^1 \cdot 3^1 \\ 
6 & 2 & 1 & 4 & 3 & 12 & 2^2 \cdot 3^1 \\
\end{array}
$$


In [ ]:
2**4*5*13**3

175760